README.txt
decoroute
=========

.. contents:

Decorator style and pattern-matching based url routing library. No framework!
Very compact. Core logic less 100 lines of code. WSGI compliant. No additional API required.

Complete example
----------------

::

    import decoroute
    
    app = decoroute.App(prefix='')
    
    # pure WSGI
    @app.expose('/wsgi/<id:\d+>')
    @decoroute.wsgi('wsgiorg.routing_args')
    def wsgi_app(environ, start_response):
        start_response('200 OK', [('Content-Type', 'text/plain')])
        return environ['wsgiorg.routing_args']
    
    # pass variables into **kw
    def render_response(status = '200 OK', content_type = 'text/plain', add_headers = [], **context):
        return status, [('Content-Type', content_type)] + add_headers, context
    
    @app.expose('/node', id = '1')
    @app.expose('/node/<id:\d+>')
    def node(env, id):
        return render_response(id = id)
    
    @app.expose('/url_for')
    def url_for(env):
        return render_response(url = app.url_for(node, id = 666))
    
    @app.expose('/404')
    def not_found(env):
        raise decoroute.NotFound()
    
    @app.not_found
    def not_found_handler(env):
        return render_response(status = '404 NF', **env)
    
    @app.render
    def render(env, (status, headers, context)):
        context['site_uri'] = '%s://%s' % (env['wsgi.url_scheme'], env['HTTP_HOST'])
        return status, headers, ['%s=%s\n' %(k, context[k]) for k in context.keys()] # fake templating
    
    from wsgiref.simple_server import make_server
    
    make_server('', 6666, app).serve_forever()

Tips
----

If you place your views (Django term. in ror - controller) in another modules, organize code like
`werkzeug <http://dev.pocoo.org/hg/werkzeug-main/file/tip/examples/shorty/>`_.

::

    # utils.py
    
    import decoroute
    app = decoroute.App()

::

    # view.py and anoter
    
    from utils import *
    
    @app.render
    def render(env, ...)
        # returns of your endpoints pass to render handler
        # in render use your prefered template engine and return triple
        # status, generator of response headers, generator of response body
        # ( generator / iterator / list - any )
        # if you use default render endpoints must be return triple
    
    @app.not_found
        # no route enpoint
        # it also pass to render handler
    
    @app.expose('/node', id = '1')
    @app.expose('/node/<id:\d+>')
    def node(env, id):
        # endpoint passed some variables
        # env is a wsgi environ
        # id = '1' default value of parameter
        #   all parameters must be <type 'str'>
        # <id:\d+> parameter pattern;
        #   \d+ specify regexp constraint of parameter (required)

::

    # manage.py - main module
    
    from utils import app
    import view # and all other
    
    if __name__ == '__main__':
        # serve app here

Legal
-----

decoroute distributed under terms of
`GNU LGPL v.2.1 <http://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt>`_.

Copyright 2008, 2009 `Vsevolod Balashov <http://vsevolod.balashov.name/>`_.

Links
-----

Source code of `decoroute <http://bitbucket.org/sevkin/decoroute/>`_.
Arch Linux `PKGBUILD <http://aur.archlinux.org/packages.php?ID=31564>`_.


debian/changelog
decoroute (0.8.1-0sg1) lucid; urgency=low

  [ Ian Eure ]
  * Update standards-version.
  * Fix.
  * Run tests & coverage in the build.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 11 Nov 2010 18:43:53 +0000

decoroute (0.8.1-0sg0) foss-development; urgency=low

  * SimpleGeoized.

 -- Ian Eure <ian@simplegeo.com>  Thu, 11 Nov 2010 10:24:11 -0800



test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# vim:ts=4:sw=4:et
# $Id: test.py,v 6357de52402e 2008/06/30 05:59:15 vsevolod $
#

from nose.tools import raises
import decoroute

class TestRE:
    
    def test_make_url_for(self):
        assert decoroute.make_url_for('/') == '/'
        assert decoroute.make_url_for('/node') == '/node'
        assert decoroute.make_url_for('/node<id:\d+>') == '/node%(id)s'
        decoroute.make_url_for('/node<id:\d+>/<name:[^>]+>') == '/node%(id)s/%(name)s'
    
    def test_make_variables(self):
        assert decoroute.make_variables('/') == []
        assert decoroute.make_variables('/node') == []
        assert decoroute.make_variables('/node<id:\d+>') == ['id']
        assert decoroute.make_variables('/node<id:\d+>/<name:[^/]+>') == ['id', 'name']
    
    def test_make_selector_fragment(self):
        assert decoroute.make_selector_fragment('/') == '\\/'
        assert decoroute.make_selector_fragment('/node') == '\\/node'
        assert decoroute.make_selector_fragment('/node<id:\d+>') == '\\/node(?:\\d+)'
        assert decoroute.make_selector_fragment('/node<id:\d+>/<name:[^/]+>') == '\\/node(?:\\d+)\\/(?:[^/]+)'
    
    def test_make_selector(self):
        assert decoroute.make_selector(['/']) == '(^\\/$)'
        assert decoroute.make_selector(['/', '/node']) == '(^\\/$)|(^\\/node$)'
        assert decoroute.make_selector(['/', '/node', '/node<id:\d+>']) == '(^\\/$)|(^\\/node$)|(^\\/node(?:\\d+)$)'

class TestUrlMap:
    
    def setUp(self):
        self.map = decoroute.UrlMap()
    
    def test_route(self):
        def p(cid, pid):
            pass
        self.map.add('/category_<cid:\d+>.html', p, pid = '1')
        self.map.add('/category_<cid:\d+>/page_<pid:\d+>.html', p)
        
        assert (p, dict(cid='1', pid='1')) == self.map.route('/category_1.html')
        assert (p, dict(cid='1', pid='1')) == self.map.route('/category_1/page_1.html')
        assert (p, dict(cid='1', pid='2')) == self.map.route('/category_1/page_2.html')
    
    def test_url_for(self):
        def p(cid, pid):
            pass
        self.map.add('/category_<cid:\d+>.html', p, pid = '1')
        self.map.add('/category_<cid:\d+>/page_<pid:\d+>.html', p)
        
        assert '/category_1.html' == self.map.url_for(p, cid='1')
        assert '/category_1.html' == self.map.url_for(p, cid='1', pid='1')
        assert '/category_1/page_2.html' == self.map.url_for(p, cid='1', pid='2')
    
    @raises(Exception)
    def test_param_defaults_subset(self):
        def ab(a, b):
            pass
        self.map.add('/a_<a:\d+>/b_<b:\d+>', ab)
        self.map.add('/static', ab, a=1, b=2)
        try:
            self.map.add('/a/b_<b:\d+>', ab, b=2)
        except Exception, e:
            assert e.args[2] == dict(b=2) and e.args[3] == dict(a=1, b=2)
            raise
    
    @raises(Exception)
    def test_param_defaults_superset(self):
        def ab(a, b):
            pass
        self.map.add('/a_<a:\d+>/b_<b:\d+>', ab)
        self.map.add('/a/b_<b:\d+>', ab, b=2)
        try:
            self.map.add('/static', ab, a=1, b=2)
        except Exception, e:
            assert e.args[2] == dict(a=1, b=2) and e.args[3] == dict(b=2)
            raise

class TestApp:
    pass

class TestRequestMethod:
    pass


decoroute.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# $Id: decoroute.py,v 7027370799fe 2009/10/28 07:01:52 vsevolod $
#
# Pattern Matching based WSGI-enabled URL routing tool.
#
# Actual version on http://pypi.python.org/pypi/decoroute
#
# (C) 2008, 2009 by Vsevolod S. Balashov <vsevolod@balashov.name>
# under terms of GNU LGPL v2.1 http://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt

__author__  = "Vsevolod Balashov"
__email__   = "vsevolod@balashov.name"
__version__ = "0.8.1"

import re
from collections import defaultdict

__all__ = ['NotFound', 'App', 'wsgi']

class NotFound(Exception):
    pass

_pattern_re = re.compile(r'''
    ([^<]+)                     # static rule data
    (?:<
        ([a-zA-Z][a-zA-Z0-9_]*) # variable name
        \:
        ([^>]+)                 # regexp constraint
    >)?
''', re.VERBOSE)

def pattern2regexp(pattern, f, s = lambda x: re.escape(x)):
    def parser():
        for t in _pattern_re.findall(pattern):
            yield s(t[0])
            if t[1] != '':
                yield f(t[1], t[2])
    return parser()

make_url_for = lambda p: ''.join(pattern2regexp(p, lambda v, r: '%%(%s)s' % v, lambda s: s))
make_variables = lambda p: filter(lambda x: x, pattern2regexp(p, lambda v, r: v, lambda s: None))
make_pattern = lambda p: r''.join(pattern2regexp(p, lambda v, r: r'(?P<%s>%s)' % (v, r)))
make_selector_fragment = lambda p: r''.join(pattern2regexp(p, lambda v, r: r'(?:%s)' % r))
make_selector = lambda i: r'(^%s$)' % r'$)|(^'.join(map(make_selector_fragment, i))

class UrlMap(object):
    "if you understand it - you are python jedi, else - just use it"
    
    def __init__(self):
        self._endpoints = {}
        self._defaults = defaultdict(lambda:{})
        self._patterns = {}
        self._pattern_selector = re.compile(make_selector(self._patterns.iterkeys()))
    
    def add(self, pattern, endpoint, **kw):
        if self._patterns.has_key(pattern):
            raise Exception('duplicate pattern', pattern)
        self._endpoints[(endpoint, frozenset(make_variables(pattern)))] = make_url_for(pattern)
        keyvalues = frozenset([(k, v) for k,v in kw.iteritems()])
        if keyvalues != frozenset():
            if self._defaults[endpoint].has_key(keyvalues):
                raise Exception('duplicate defaults for endpoint', endpoint, kw)
            for key in self._defaults[endpoint].keys():
                if keyvalues.issuperset(key) or keyvalues.issubset(key):
                    raise Exception('defaults for endpoint already exist in route', endpoint, \
                      kw, dict([c for c in key]))
            self._defaults[endpoint][keyvalues] = set(kw.keys())
        self._patterns[pattern] = (re.compile(make_pattern(pattern)), endpoint, kw)
        self._pattern_selector = re.compile(make_selector(self._patterns.iterkeys()))
    
    def route(self, url):
        """return (endpoint, kw)
            kw = parsed from url + defaults for this url
        """
        try:
            p = self._patterns.values()[re.match(self._pattern_selector, url).lastindex - 1]
            m = p[0].match(url)
            d = m.groupdict().copy()
            d.update(p[2])
            # l = list(m.groups()) ## http://wsgi.org/wsgi/Specifications/routing_args#example
            # e = url[m.end():]
            return p[1], d #, l, e
        except Exception, e:
            raise NotFound('route not found', url)
    
    def url_for(self, endpoint, **kw):
        keyvalues = frozenset([(k, v) for k,v in kw.iteritems()])
        keys = set(kw.keys())
        for defaults in self._defaults[endpoint].keys():
            if defaults.issubset(keyvalues):
                keys -= self._defaults[endpoint][defaults]
                #TODO what if find another?
                break
        return self._endpoints[(endpoint, frozenset(keys))] % kw

class App(object):
    def __init__(self, prefix = '', key = 'decoroute.app'):
        self.map = UrlMap()
        self._prefix = (prefix, re.compile(r'^%s' % re.escape(prefix)))
        self._key = key
        self._not_found = lambda e: ('404 NOT FOUND', [("Content-Type", "text/plain")], [''])
        self._render = lambda e, r: r
    
    def route(self, env):
        env[self._key] = self
        path, n = self._prefix[1].subn('', env['PATH_INFO'])
        if n == 1:
            endpoint, kw = self.map.route(path)
        else:
            raise NotFound()
        return endpoint(env, **kw)
    
    def url_for(self, endpoint, **kw):
        return self._prefix[0] + self.map.url_for(endpoint, **kw)
    
    def __call__(self, env, start_response):
        def helper((response, headers, body)):
            start_response(response, headers)
            return body
        try:
            return helper(self._render(env, self.route(env)))
        except NotFound:
            return helper(self._render(env, self._not_found(env)))
    
    def expose(self, pattern, **kw):
        def decorate(f):
            self.map.add(pattern, f, **kw)
            return f
        return decorate
    
    def not_found(self, f):
        self._not_found = f
        return f
    
    def render(self, f):
        self._render = f
        return f

class Response(object):
    def __call__(self, status, headers):
        self.status = status
        self.headers = headers

def wsgi(key_name):
    "put kw into env[key_name] and make endpoint wsgi-compliant"
    def decorate(f):
        def fun(env, **kw):
            response = Response()
            env[key_name]=kw
            body=f(env, response)
            return response.status, response.headers, body
        return fun
    return decorate

def request_method(*a):
    "not implemented"
    def decorate(f):
        s = frozenset(a)
        def proxy(env, **kw):
            assert s.issuperset(frozenset([env['REQUEST_METHOD']]))
            return f(env, **kw)
        return proxy # , s
    return decorate


demo.py
## $Id: demo.py,v c9e97604c73e 2008/07/30 11:26:55 vsevolod $

import decoroute
    
app = decoroute.App(prefix='')
    
# pure WSGI
@app.expose('/wsgi/<id:\d+>')
@decoroute.wsgi('path.kw')
def wsgi_app(environ, start_response):
    start_response('200 OK', [('Content-Type', 'text/plain')])
    return environ['path.kw']
    
# pass variables into **kw
def render_response(status = '200 OK', content_type = 'text/plain', add_headers = [], **context):
    return status, [('Content-Type', content_type)] + add_headers, context
    
@app.expose('/node', id = '1')
@app.expose('/node/<id:\d+>')
def node(env, id):
    return render_response(id = id)
    
@app.expose('/url_for')
def url_for(env):
    return render_response(url = app.url_for(node, id = 666))
    
@app.expose('/404')
def not_found(env):
    raise decoroute.NotFound()
    
@app.not_found
def not_found_handler(env):
    return render_response(status = '404 NF', **env)
    
@app.render
def render(env, (status, headers, context)):
    context['site_uri'] = '%s://%s' % (env['wsgi.url_scheme'], env['HTTP_HOST'])
    return status, headers, ['%s=%s\n' %(k, context[k]) for k in context.keys()] # fake templating
    
from wsgiref.simple_server import make_server
    
make_server('', 6666, app).serve_forever()
